BUG: Fix a bug when using DataFrame.to_records with unicode column names

AlexisMignon opened the pull request:
Fix a bug when using DataFrame.to_records with unicode column names in python 2
- [x] closes #11879
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

AlexisMignon committed:
BUG: Fix a bug occuring when using DataFrame.to_records with unicode
     column names in python 2. (issue #11879)

jreback commented:
test please

AlexisMignon committed:
BUG: Add a non regression test for to_records() with unicode column names in python 2

sinhrks commented on pandas/tests/frame/test_convert_to.py:
```diff
@@ -179,3 +179,11 @@ def test_to_records_with_unicode_index(self):
+    def test_to_records_with_unicode_column_names(self):
+        # Issue #11879. to_records used to raise an exception when used
+        # with column names containing non ascii caracters in Python 2
+        try:
```
Isn't this being fixed by this PR? If so, remove try-except block and compare the result with the expected result.

AlexisMignon replied:
The expected result is that it doesn't raise an exception.

sinhrks replied:
Then, you can remove try-except and compare the result with expected array.  We don't use ``self.fail`` often because any error results in failure.

jreback commented:
can you rebase / update?

AlexisMignon committed:
Resyncing pull request with pydata/master

AlexisMignon commented:
I've resynced my branch but it still fails. Not sure whether it is specifically due to my code though.

jreback commented:
can you rebase / update according to comments

AlexisMignon committed:
Resyncing fork with pydata/pandas

jreback commented:
closing as stale. if you want to update, pls comment.

AlexisMignon committed:
Changed the way dtype is specified in to_records

Changed the way dtype is specified in to_records in order to allow unicode field names.

AlexisMignon committed:
BUG: Reformatted the test as requested

AlexisMignon commented:
I changed the test as requested. Note that i corrected another problem due to the fact that numpy does not allow to specify dtype with unicode field names as list of tuples, but allows it using dictionnaries

AlexisMignon committed:
[BUG] Added forgotten assert in test

AlexisMignon committed:
CLN: Make the test complient with flake8

AlexisMignon committed:
CLN: Remove flake8 warnings

jreback commented on pandas/tests/frame/test_convert_to.py:
```diff
@@ -179,3 +179,16 @@ def test_to_records_with_unicode_index(self):
+        # with column names containing non ascii caracters in Python 2
+        result = DataFrame(data={u"accented_name_é": [1.0]}).to_records()
+        # Note that numpy allows for unicode field names but dtypes need
+        # to be specified using dictionnary intsead of list of tuples.
```
do you have a reference for this? is it listed as a numpy bug? (if not it should be)

AlexisMignon replied:
This is referenced here:
https://github.com/numpy/numpy/issues/2407

jreback commented:
pls add a whatsnew for 0.20.0. lgtm. otherwise

AlexisMignon committed:
Added bug fix entry in what's new

AlexisMignon committed:
Merge branch 'master' into master

jreback commented:
can you rebase / update

jreback commented:
thanks @AlexisMignon
